src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -442,37 +470,92 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
+            // remove the prefix of "$AWS/rules/rule-name/"
+            topic = topic.split(prefixOfReservedTopic, 2)[1];
+        }
+
+        // Topic should not have no more than maximum number of forward slashes (/)
+        if (topic.chars().filter(num -> num == '/').count() > DEFAULT_MAXIMUM_NUMBER_OF_FORWARD_SLASHES) {
+            logger.atInfo().kv("topic", topic).log("new topic - for /");
+            String errMsg = String.format("The topic of publish request must have no"
+                    + "more than %d forward slashes (/). This excludes the first 3 slashes in the mandatory segments "
+                    + "for Basic Ingest topics ($AWS/rules/rule-name/).", DEFAULT_MAXIMUM_NUMBER_OF_FORWARD_SLASHES);
+            throw new SpoolerStoreException(errMsg);
+        }
+
+        // Check the topic size
+        if (topic.getBytes(StandardCharsets.UTF_8).length > DEFAULT_MAXIMUM_LENGTH_OF_TOPIC) {
+            logger.atInfo().kv("topic", topic).log("new topic - for size");
+            String errMsg = String.format("The topic size of publish request must be no "
+                            + "larger than {} bytes of UTF-8 encoded characters. This excludes the first "
+                            + "3 mandatory segments for Basic Ingest topics ($AWS/rules/rule-name/).",
+                    DEFAULT_MAXIMUM_LENGTH_OF_TOPIC);
+            throw new SpoolerStoreException(errMsg);
+        }
     }
 
     @SuppressWarnings({"PMD.AvoidCatchingThrowable", "PMD.PreserveStackTrace"})
     protected CompletableFuture<Integer> publishSingleSpoolerMessage() throws InterruptedException {
         long id = -1L;
         try {
             id = spool.popId();
-            PublishRequest request = spool.getMessageById(id);
+            SpoolMessage spooledMessage = spool.getMessageById(id);
+            PublishRequest request = spooledMessage.getRequest();
             MqttMessage m = new MqttMessage(request.getTopic(), request.getPayload());
 
             long finalId = id;
             return getConnection(false).publish(m, request.getQos(), request.isRetain())
                     .whenComplete((packetId, throwable) -> {
                         // packetId is the SDK assigned ID. Ignore this and instead use the spooler ID
                         if (throwable == null) {
+                            spooledMessage.getFuture().complete(packetId);
                             spool.removeMessageById(finalId);
                             logger.atDebug().kv("id", finalId).kv("topic", request.getTopic())
                                     .log("Successfully published message");
                         } else {
-                            spool.addId(finalId);
-                            logger.atError().log("Failed to publish the message via Spooler and will retry",
-                                    throwable);
+                            if (spooledMessage.getRetried().get()
```

Comment:
Just use `getAndIncrement()` right here so that it is automatically incremented and you don't need to do it on L551